touch/touch.1
.\"
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)touch.1	1.18 (gritter) 2/10/05
.TH TOUCH 1 "2/10/05" "Heirloom Toolchest" "User Commands"
.SH NAME
touch \- update file access and modification times
.SH SYNOPSIS
\fBtouch\fR [\fB\-amc\fR] [\fIMMDDhhmm\fR[\fIYY\fR]] \fIfile\fR ...
.br
\fBtouch\fR [\fB\-amc\fR]
[\fB\-r\fI\ ref_file\fR|\fB\-t\fI\ time\fR] \fIfile\fR ...
.SH DESCRIPTION
The
.B touch
utility
changes the access and modification times of files.
It creates a file if it does not already exist.
If none of the arguments specify the time to set,
the current time is used.
.PP
If neither the
.B \-r
nor the
.B \-t
option is used to specify a time value
and the first non-option argument begins with a digit,
and it consists of exactly eight or ten digits with
.B /usr/5bin/posix/touch
and
.BR /usr/5bin/posix2001/touch ,
the times of the files are set to the given time.
.PP
The meaning of fields in time specifications is as follows:
.RS
.PP
\fICC\fP	The first two digits of the year.
.br
\fIYY\fP	The second two digits of the year.
.br
\fIMM\fP	The month of the year (01-12).
.br
\fIDD\fP	The day of the month (01-31).
.br
\fIhh\fP	The hour of the day (00-23).
.br
\fImm\fP	The minute of the hour (00-59).
.br
\fISS\fP	The second of the minute (00-61).
.RE
.PP
Unless
.I YY
is present, the current year is used.
If
.I YY
is given without
.IR CC ,
the century is 19 if
.I YY
is in the range of 69\^\(en\^99,
or 20 for
.I YY
values of 00\^\(en\^68.
.PP
The following options are accepted:
.TP
.B \-a
Change the access time,
but not the modification time unless
.B \-m
is also present.
.TP
.B \-c
If a file does not exist, it is not created.
No diagnostic message is printed; for
.B /usr/5bin/touch
and
.BR /usr/5bin/s42/touch ,
the exit status is incremented.
The exit status is not affected for
.B /usr/5bin/posix/touch
and
.BR /usr/5bin/posix2001/touch .
.TP
.B \-m
Change the access time,
but not the modification time unless
.B \-a
is also present.
.PP
The following options have been introduced by POSIX.2:
.TP
.BI \-r \ ref_file
The corresponding time of the specified file is used
instead of the current time.
.TP
.BI \-t \ time
Instead of the current time,
the
.I time
specified as
.RS
.IP
.RI [[ CC ] YY ] MMDDhhmm [. SS ]
.PP
is used.
.RE
.PP
The exit status is the number of files
that were not successfully processed.
.SH "SEE ALSO"
date(1),
utime(2)


CHANGES
Release ...
* seq: Made its payload similar to UNIX v8's implementation, but with some
  extensions such as a fast algorithm for -w or -s --- which was already
  present before the (almost) full rewrite.
* timeout: avoid buffer overrun, also did some changes at the source code,
  such as cleaning up comments and initializing 'commandv' and 'fst_commandv'.
* watch: Used esr's deheader for removing unnecessary headers, also use
  pfmt(stderr, MM_ERROR, ...)  instead of prerror().
* watch, timeout: Some small corrections at the manual page.
* cpio: Skip over zip extra headers with fewer than four bytes (the minimum
  length for them to contain any information) instead of reporting a bad
  header and terminating.
* pg: Handle cases where standard output is a terminal but not open for
  reading (bug reported by Keith Doyle).
* cpio, tar: If an archive contains a file that is a hardlink to itself on
  extraction, do nothing at all instead of discarding its data. This mainly
  happens if a file with more than one link is included under the same name
  in an archive twice. Since special situations on extraction such as bind
  mounts on Linux may lead to this scenario even in cases where both link
  names are different, this is now treated like a successful link() step
  (bugreport by Thomas Graf).
* Made the getopt() substitute work with the getopt$UNIX2003() function
  redirection mechanism on Mac OS X Leopard.
* mknod, pgrep, ps: Mac OS X Leopard compilation fixes.
* w, whodo: On a 32-bit system where /proc/kcore is larger than 2 GB,
  avoid a harmless error message (bug reported by Pierre Gentile).
* ptime: Works on recent Linux kernels again.
* col: The behavior of the -x flag has been inverted to actually match
  the documentation, i.e. to suppress conversion of spaces to tabs (patch
  by Markus Schnalke).
* cut: With "-s", print the last field of a line that was also the last
  field specified with "-f" (bugreport by Pierre Gentile).

Release 240220
* [NEW] seq: Stone-simple implementation; needs some improvement yet.
* [NEW] watch: Complete, simple and stable implementation.
* [NEW] timeout: Complete implementation with extended payloads.
* readlink: Different help messages for both UCB and default variants. 

Release 230503 
* chroot: "docs: Update date at chroot.8 manual page."
* chroot: "chore: Remove comment, will move to GitHub"
* chroot: "docs: Rewrote chroot(8) manual page"
* chroot: "chore: Add ``usage()`` function at chroot.c"
* tabs: "fix: [t,g]roff formatting at tabs/fspec.5."
* readlink: "docs: Created new manual page for the non-UCB variant. (#24)"
* Makefile.mk: "fix: Don't symbolic/hard link readlink from UCBDIR to DEFDIR anymore"
* readlink: "docs: Add comments clarifying the C preprocessor directives."
* readlink: "fix: Ignore ``-n`` option when operating multiple files."
* readlink: "feat: Create non-UCB/GNU-alike variant."
* readlink: "fix: Made ``name_size`` private again, so ``pfmt()`` won't go segmentation fault."
* readlink: "feat: First step, start modularizing the code."
* libcommon: "fix: getdirentries replacement for OBSD >5.4"
* fix-me: For new releases, finish porting for OpenBSD and implement watch(1) and timeout(1).

Release 230305
* getconf: "Now _NPROCESSORS_ONLN and _NPROCESSORS_CONF are implemented."
* (*): "Removed the remaining licence headers, now every licence is referenced by its SPDX identifier."

Release 230228
* [NEW] readlink: New implementation, extensively based on OpenBSD's.
* readlink: It's installed at /usr/ucb per default and symbolic/hard linked to /usr/5bin. 

Release 230120
* Notes for the library directories if linking against the GNU C library.
* rm: Consistency within standards: the SUS program shouldn't have the POSIX 1003.1(2008) fix, just the main program.
* Made $(ROOT) not hardcoded anymore.
* Cleaned up more licence headers (if my estimations are correct, we've got rid of 3.255 lines!).
* Small update on README.md --- although it's now less readable on plain-text readers such as more or pg.

Release 230107
* rm: Added support for the POSIX 1003.1(2008).
* chroot: Added manual page.
* awk: Added stdint.h to awk.g.y.
* Now it has a bigger range of compatibility with GNU C library versions.

Release 221025
* Made notes clearer about building Heirloom on modern versions of the GNU C library.
* tar: Support for zstd and xz.

Release 220821
* [NEW] chroot: New implementation of chroot, based on Solaris 2.5 payloads.
* true, false: Rewritten in C.
* yes: Rewritten and adapted to ANSI C.
* "libcommon/vpfmt.c": Adapted to ANSI C.
* Now many of the utilities can be built in the musl C library and GNU C library (dynamically).

Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.

touch/touch.c
/*
 * touch - update access and modification times of a file
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, July 2001.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#if defined (SUS)
static const char sccsid[] USED = "@(#)touch_sus.sl	1.21 (gritter) 5/29/05";
#else	/* !SUS */
static const char sccsid[] USED = "@(#)touch.sl	1.21 (gritter) 5/29/05";
#endif	/* !SUS */

#include	<sys/types.h>
#include	<sys/stat.h>
#include	<fcntl.h>
#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<utime.h>
#include	<ctype.h>
#include	<time.h>

static unsigned	errcnt;			/* count of errors */
static int	aflag;			/* select access time */
static int	cflag;			/* do not create files */
static int	mflag;			/* select modification time */
static int	settime;		/* this is the settime command */
static int	date_time;		/* use date_time operand (not -t) */
static time_t	nacc = -1;		/* new access time */
static time_t	nmod = -1;		/* now modification time */
static time_t	now;			/* current time */
static char	*progname;		/* argv[0] to main() */
static int	nulltime;		/* can use settime(NULL) */

/*
 * perror()-alike.
 */
static void
pnerror(int eno, const char *string)
{
	fprintf(stderr, "%s: %s: %s\n", progname, string, strerror(eno));
	errcnt++;
}

/*
 * Touch the named file.
 */
static void
touch(const char *fn)
{
	struct stat st;
	struct utimbuf ut;

	if (stat(fn, &st) < 0) {
		if (errno == ENOENT) {
			int fd;

			if (cflag) {
#ifndef	SUS
				errcnt++;
#endif	/* !SUS */
				return;
			}
			if ((fd = creat(fn, 0666)) < 0) {
				fprintf(stderr, "%s: %s cannot create\n",
						progname, fn);
				errcnt++;
				return;
			}
			if (fstat(fd, &st) < 0) {
				fprintf(stderr, "%s: %s cannot stat\n",
						progname, fn);
				errcnt++;
				close(fd);
				return;
			}
			close(fd);
		} else {
			fprintf(stderr, "%s: %s cannot stat\n",
					progname, fn);
			errcnt++;
			return;
		}
	}
	if (aflag)
		ut.actime = nacc;
	else
		ut.actime = st.st_atime;
	if (mflag)
		ut.modtime = nmod;
	else
		ut.modtime = st.st_mtime;
	if (utime(fn, nulltime ? NULL : &ut) < 0) {
		fprintf(stderr, "%s: cannot change times on %s\n",
				progname, fn);
		errcnt++;
	}
}

static void
badtime(void)
{
	if (date_time)
		fprintf(stderr, "date: bad conversion\n");
	else
		fprintf(stderr, "%s: bad time specification\n", progname);
	exit(2);
}

/*
 * Convert s to int with maximum of m.
 */
static int
atot(const char *s, int m)
{
	char *x;
	int i;

	i = (int)strtol(s, &x, 10);
	if (*x != '\0' || i > m || i < 0 || *s == '+' || *s == '-')
		badtime();
	return i;
}

/*
 * Interpret a time argument, old style: MMDDhhmm[YY].
 */
static void
otime(char *cp)
{
	struct tm *stm;
	time_t t;

	date_time = 1;
	t = now;
	stm = localtime(&t);
	stm->tm_isdst = -1;
	stm->tm_sec = 0;
	switch (strlen(cp)) {
	case 10:
		stm->tm_year = atot(&cp[8], 99);
		if (stm->tm_year < 69)
			stm->tm_year += 100;
		cp[8] = '\0';
		/*FALLTHRU*/
	case 8:
		stm->tm_min = atot(&cp[6], 59);
		cp[6] = '\0';
		stm->tm_hour = atot(&cp[4], 23);
		cp[4] = '\0';
		if ((stm->tm_mday = atot(&cp[2], 31)) == 0)
			badtime();
		cp[2] = '\0';
		if ((stm->tm_mon = atot(cp, 12)) == 0)
			badtime();
		stm->tm_mon--;
		break;
	default:
		badtime();
	}
	if ((t = mktime(stm)) == (time_t)-1)
		badtime();
	nacc = nmod = t;
}

/*
 * Interpret a time argument, new style: [[CC]YY]MMDDhhmm[.SS].
 */
static void
ptime(char *cp)
{
	char year[5];
	struct tm *stm;
	time_t t;
	size_t sz = strlen(cp);

	t = now;
	stm = localtime(&t);
	stm->tm_isdst = -1;
	if (sz == 11 || sz == 13 || sz == 15) {
		if (cp[sz - 3] != '.')
			badtime();
		stm->tm_sec = atot(&cp[sz - 2], 61);
		cp[sz - 3] = '\0';
		sz -= 3;
	} else
		stm->tm_sec = 0;
	if (sz == 12) {
		year[0] = cp[0], year[1] = cp[1], year[2] = cp[2],
			year[3] = cp[3], year[4] = '\0';
		if ((stm->tm_year = atot(year, 30000)) < 1970)
			badtime();
		stm->tm_year -= 1900;
		cp += 4, sz -= 4;
	} else if (sz == 10) {
		year[0] = cp[0], year[1] = cp[1], year[2] = '\0';
		stm->tm_year = atot(year, 99);
		if (stm->tm_year < 69)
			stm->tm_year += 100;
		cp += 2, sz -= 2;
	}
	if (sz != 8)
		badtime();
	stm->tm_min = atot(&cp[6], 59);
	cp[6] = '\0';
	stm->tm_hour = atot(&cp[4], 23);
	cp[4] = '\0';
	if ((stm->tm_mday = atot(&cp[2], 31)) == 0)
		badtime();
	cp[2] = '\0';
	if ((stm->tm_mon = atot(cp, 12)) == 0)
		badtime();
	stm->tm_mon--;
	if ((t = mktime(stm)) == (time_t)-1)
		badtime();
	nacc = nmod = t;
}

/*
 * Get reference time from a file.
 */
static void
reffile(const char *cp)
{
	struct stat st;

	if (stat(cp, &st) < 0) {
		pnerror(errno, cp);
		exit(1);
	}
	nacc = st.st_atime;
	nmod = st.st_mtime;
}

static void
usage(void)
{
	if (settime == 0)
		fprintf(stderr, "usage: %s [-amc] [mmddhhmm[yy]] file ...\n",
       			progname);
	else
		fprintf(stderr, "usage: %s [-f file] [mmddhhmm[yy]] file ...\n",
				progname);
	exit(2);
}

int
main(int argc, char **argv)
{
	int i;

#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	progname = basename(argv[0]);
	if (strcmp(progname, "settime") == 0) {
		settime = 1;
		cflag = 1;
	}
	time(&now);
	while ((i = getopt(argc, argv, settime ? "f:" : "amcr:t:f")) != EOF) {
		switch (i) {
		case 'a':
			aflag = 1;
			break;
		case 'c':
			cflag = 1;
			break;
		case 'm':
			mflag = 1;
			break;
		case 'f':
			if (settime == 0)
				break;
			/*FALLTHRU*/
		case 'r':
			if (nacc != (time_t)-1 || nmod != (time_t)-1)
				usage();
			reffile(optarg);
			break;
		case 't':
			if (nacc != (time_t)-1 || nmod != (time_t)-1)
				usage();
			ptime(optarg);
			break;
		default:
			usage();
		}
	}
	if (nacc == (time_t)-1 && nmod == (time_t)-1 && argv[optind]
#ifdef	SUS
			&& argv[optind + 1]) {
		char	*cp;
		for (cp = argv[optind]; *cp && isdigit(*cp & 0377); cp++);
		if (*cp == '\0' && (cp - argv[optind] == 8 ||
					cp - argv[optind] == 10))
#else	/* !SUS */
			) {
		if (isdigit(argv[optind][0]))
#endif	/* !SUS */
			otime(argv[optind++]);
	}
	if (nacc == (time_t)-1 && nmod == (time_t)-1 && aflag == 0 &&
			mflag == 0)
		nulltime = 1;
	if (nacc == (time_t)-1)
		nacc = now;
	if (nmod == (time_t)-1)
		nmod = now;
	if (aflag == 0 && mflag == 0)
		aflag = mflag = 1;
#ifdef	SUS
	if (optind >= argc)
		usage();
#else	/* !SUS */
	if (optind >= argc && date_time == 0)
		usage();
#endif	/* !SUS */
	for (i = optind; i < argc; i++)
		touch(argv[i]);
	return errcnt < 0100 ? errcnt : 077;
}
